src/backend/distributed/commands/utility_hook.c

```diff
@@ -1627,17 +1664,93 @@ RunPreprocessMainDBCommand(Node *parsetree, const char *queryString)
 static void
 RunPostprocessMainDBCommand(Node *parsetree)
 {
+	if (IsStatementSupportedInNonMainDb(parsetree) &&
+		StatementRequiresMarkDistributedFromNonMainDb(parsetree))
+	{
+		ObjectInfo objectInfo = GetObjectInfo(parsetree);
+		MarkObjectDistributedInNonMainDb(parsetree, objectInfo);
+	}
+}
+
+
+/*
+ * GetObjectInfo returns the name and oid of the object in the given parsetree.
+ */
+static ObjectInfo
+GetObjectInfo(Node *parsetree)
+{
+	ObjectInfo info;
+
 	if (IsA(parsetree, CreateRoleStmt))
 	{
-		StringInfo mainDBQuery = makeStringInfo();
-		CreateRoleStmt *createRoleStmt = castNode(CreateRoleStmt, parsetree);
-		Oid roleOid = get_role_oid(createRoleStmt->role, false);
-		appendStringInfo(mainDBQuery,
-						 MARK_OBJECT_DISTRIBUTED,
-						 AuthIdRelationId,
-						 quote_literal_cstr(createRoleStmt->role),
-						 roleOid,
-						 quote_literal_cstr(CurrentUserName()));
-		RunCitusMainDBQuery(mainDBQuery->data);
+		CreateRoleStmt *stmt = castNode(CreateRoleStmt, parsetree);
+		info.name = stmt->role;
+		info.id = get_role_oid(stmt->role, false);
 	}
+
+	/* Add else if branches for other statement types */
+
+	return info;
+}
```

Comment:
Renaming ObjectInfo to MarkObjectDistributedParams to make it clear what it's about and then ..

```c
/*
 * GetMarkObjectDistributedParams returns MarkObjectDistributedParams for the object
 * that needs to be marked as distributed when executing given parsetree from a
 * non-main database.
 */
static MarkObjectDistributedParams
GetMarkObjectDistributedParams(Node *parsetree)
{
	ObjectInfo info;

	if (IsA(parsetree, CreateRoleStmt))
	{
		CreateRoleStmt *stmt = castNode(CreateRoleStmt, parsetree);
		info.name = stmt->role;
		bool missingOk = false;
		info.id = get_role_oid(stmt->role, missingOk);
		info.catalogRelId = AuthIdRelationId;
	}

	/* Add else if branches for other statement types */

	return info;
}
```

Reply:
The aim of this method is not limited to MarkDistributedParameters. This method extracts oid and name parameters, which is pretty common for nearly all statements. I even needed to change the structure to handle multiple values to use it for Unmark operation. I prefer keeping as it is. I can change the name ExtractObjectDetailsFromNode or GetObjectDetailsFromNode.

Reply:
For reference of the further usage https://github.com/citusdata/citus/pull/7461#discussion_r1473998572

Reply:
Current naming seems fine then. And this is another why we should move all those logic into a separate file in a separate PR because `ObjectInfo` doesn't sound like a struct that is useful for whole utility hook but useful for non-main database code-paths, anyway.

Reply:
Renamed as you suggested